### DISCUSSION TITLE
Add Postgresql

### DISCUSSION THREAD

**Discussion by hopkins385** [serversideup/docker-php#158] [Ideas & Feature Requests]
### Problem
Hi guys, I was trying to use Postgresql, but it's not available within this docker images.

### Impact
I think providing the flexibility to select any SQL driver which is supported by Laravel (default) can be beneficial for all of us. 

### Solution
`apt-get install postgresql`

Reference: [https://www.postgresql.org/download/linux/debian/](https://www.postgresql.org/download/linux/debian/)

php.ini
```
extension=pdo_pgsql
extension=pgsql
```

**Comment by jaydrogers** [comment]
You should be able to extend the image and customize it to your own liking. I'm actually working on a documentation site on how to do this and will hopefully have this documentation available soon (https://feature-v2-2-0-documentation.docker-php.pages.dev/).

If I understand correctly, `apt-get install postgresql` will install Postgres Server within the container? If so, I would advise running it in a separate container as possible. Only in a few cases I would recommend running everything in a single container.

**Reply by hopkins385** [reply]
The new site looks hot! 
Agree, the postgresql server should be in a separate container. 

I tried already to change the php.ini, but haven't been able to map it to a local file or to manually edit it as vim or nano is not installed and cannot be installed afterwards.

If I would be able to change the image on my own that would also be good enough.

**Reply by jaydrogers** [reply]
Thanks!!

I actually just soft launched the new docs site. This might be a good test 😃

I wouldn't call my docs "done done", but at least I have them organized from before. I might clean them up as I go.

Check this section out: https://serversideup.net/open-source/docker-php/docs/guide/customizing-the-image

Do you know what client you need to install? I have a feeling `postgesql` will install the actual server, correct?

**Reply by hopkins385** [reply]
Hi, sorry for late reply. 

What do you mean with client?

Correct, `apt install postgresql` will install the server itself.

Maybe I havent communicated clear enough what I'm planning: 
So my plan is to activate the php postgresql extension, then to spin up a postgresql server in a separate container, and all that together I want to use in a laravel project with the native orm database driver for postgre.

**Reply by jaydrogers** [reply]
Understood!

So what you're looking for is following something like this: https://serversideup.net/open-source/docker-php/docs/guide/customizing-the-image

Here is what your `Dockerfile` will look like:

```Dockerfile
# Choose any image you want
FROM serversideup/php:8.2-fpm-nginx as base

# Install the Postgres **CLIENT** only in our PHP image
RUN apt-get update \
    && apt-get install -y --no-install-recommends postgresql-client \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/* /usr/share/doc/*

# Copy our Laravel app into the Docker image
FROM base as deploy
COPY --chown=9999:9999 . /var/www/html
```

You can remove the last `COPY` command if you don't want your app files in the image (you can use a volume to mount them instead).

Hope this helps!!

**Reply by hopkins385** [reply]
Cool, thanks for the tutorial. But how can I change the php ini file? I tried to mount it but it didn't work.

**Reply by jaydrogers** [reply]
Absolutely!

Could you give an example of what options you are trying to change in your PHP ini?

**Reply by hopkins385** [reply]
In the php Ini I want to uncomment these lines 

'''

extension=pdo_pgsql
extension=pgsql
'''

**Reply by hopkins385** [reply]
ok, so what I've figured out is that `apt-get install php-pgsql` needs to be installed. 
This adds two more php.ini's and installs the required extension.

```ini
20-pgsql.ini,
20-pdo_pgsql.ini
```

The postgres client is optional, but recommended if a manual interaction via cli is required.

So I ended up with the following:

```
# Choose any image you want
FROM serversideup/php:8.2-fpm as base

# Install the Postgres **CLIENT** only in our PHP image
RUN apt-get update \
    && apt-get install -y --no-install-recommends postgresql-client php-pgsql \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/* /usr/share/doc/*

# Copy our Laravel app into the Docker image
FROM base as deploy
```

So my question is now, how about to add the php-pgsql extension as default into the official serversideup image for php and for php-nginx? By doing so, this image would support all the SQL drivers that are "native" in Laravel and users get more flexibility in choosing their favourite sql driver.

If you ask me, I think its sufficient to just add the `php-pgsql` alone without the pgsql client.

```
&& apt-get install -y --no-install-recommends php-pgsql \
```

What do you think?

**Reply by jaydrogers** [reply]
Thanks for posting your solution! Sorry I wasn't available towards end of last week -- it was a busy one 😃

**Regarding adding `php-pgsql` by default**
I'm going to re-open this and see if there is more traction on it. I have a feeling majority of users are using MySQL-based databases, so that's why that's included by default.

Postgres is a great solution too, but I also want to keep these images as small as possible.

By leaving this thread open, others can chime in to increase demand for the feature and it will also be a note for me to investigate and see how much larger the images become from the dependency `php-pgsql` installs.

**Comment by anvme** [comment]
hey! We need php pgsql module installed by default =)
